internal/controllers/topology/cluster/current_state.go

```diff
@@ -158,6 +158,12 @@ func (r *Reconciler) getCurrentMachineDeploymentState(ctx context.Context, clust
 	for i := range md.Items {
 		m := &md.Items[i]
 
+		// check that the MachineDeployment has the ClusterTopologyOwnedLabel label.
+		// Nb. This is to make sure that a managed topology cluster does not have a reference to an object that is not
+		// owned by the topology.
+		if !labels.IsTopologyOwned(m) {
```

Comment:
If I'm reading the issue right we should also do this for the control plane templates - do you plan to add that in this PR?

Reply:
I thinks its already there for control plane : https://github.com/kubernetes-sigs/cluster-api/blob/8d7f010a3d37caa5d5f620b99976901c76c52046/internal/controllers/topology/cluster/current_state.go#L104

Reply:
Apologies - I mean ControlPlane InfrastructureMachineTemplate https://github.com/kubernetes-sigs/cluster-api/blob/8d7f010a3d37caa5d5f620b99976901c76c52046/internal/controllers/topology/cluster/current_state.go#L118

Reply:
I will add that.

Reply:
I don't think this is necessary, given that we are readying only md with this label, however, it could be interesting doing this check on bootstrap/inframachines referenced by the md

Reply:
I think we should additionally add this for MD BootstrapConfigTemplate and InfrastructureMachineTemplate (~l.196 + l.206)

This way we make sure to never treat objects as part of the topology which are not owned by us (aka the topology controller)

Reply:
@adilGhaffarDev can we drop this test on MD given that it is not necessary?

Reply:
Ah good point, missed that

Reply:
oh I forgot, I will remove it.